Init dashboard

pva701 opened the pull request:
## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-30

What it left (will be done in next pr):
- [ ] refresh subexperiments on selector change
- [ ] add context menu
- [x] add right click contenxt menu for experiment
- [ ] fix bug on removing all selected subexperiments
- [ ] hideable new subexperiment plate
- [ ] tooltip for Recompute (and for n more hover as well, and more numbers after point for IC50)
- [ ] update isSuspicious flag
- [ ] show error notify on methodology/subexperiment delete 

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

pva701 committed:
Implement Dashboard selects

Problem: it should be possible to select project, compound and target
on Dashboard page and see some information about selected item.

Solution: combine react-select with information plate.

pva701 committed:
Add experiments table

Problem: list of filtered experiments should be shown on Dahsboard page.

Solution: add table for this.

pva701 committed:
Add basic plot

Problem: selected subexperiments should be shown on chart.

Solution: use basic plotly functionality for this.

pva701 committed:
Implement collapsible experiments

Problem: experiments are visible on the table, but not its subexperiments.

Solution: add new component to show every subexperiment inside of experiment.
Make it collapsible on experiment click.

pva701 committed:
Improve UX (reset states and separate spinners)

Problem: for now when we go between pages the state is not reset,
also there is a common spinner for chart and experiments table,
which is triggered whenever we add new subexperiment on the chart.

Solution: reset states when go between pages.
Also separate spinner for two.

pva701 committed:
Fix methodology deletion dialog

Problem: when methodology deleted,
it's not immediatelly reflected on the methodologies tab.

Solution: fix async request on delete.

pva701 committed:
Fix bug: update filtered experiments when new experiments uploaded

pva701 committed:
Show basic charts

Problem: we need to show points for several subexperiments,
also 4PL for them,
using different markers for different points.

Solution: use different traces in plotly to differ them.

pva701 committed:
Make possible enable and disable points

Problem: it's not possible to enable and disabled points from subexperiment.

Solution: add new plotly trace with markers for selected points.

pva701 committed:
Move Dashboard components to separate folders

Problem: it's getting too messy in page/dashboard.

Solution: move components to their own folders.

sashasashasasha151 commented on frontend/src/pages/dashboard/ExperimentsTable/ExperimentsTable.tsx:
```diff
@@ -0,0 +1,201 @@
+            <ExpandMinimizeButton targetSize="expanded" />
+          </div>
+          <div className="tableContainer experimentsArea__experimentsTableContainer">
+            <Table<Experiment>
+              small
+              columns={expTableSize === "minimized" ? minimizedColumns : expandedColumns}
+              data={
+                showEntries === "all"
+                  ? experiments
+                  : experiments.filter(e => selectedExperiments.has(e.id))
+              }
+              collapsible={e => (
+                <SuspenseSpinner>
+                  <ExperimentsCollapse experiment={e} />
+                </SuspenseSpinner>
+              )}
+            />
```
![image](https://user-images.githubusercontent.com/25207759/114409811-82bfef00-9bb3-11eb-85f3-1aa51cac21e8.png)

pva701 replied:
Fixed in this PR

sashasashasasha151 commented on frontend/src/pages/dashboard/DashboardPage.scss:
```diff
@@ -0,0 +1,42 @@
+.dashboardPage {
+  display: grid;
+  grid-template-rows: (170 * $px) auto;
+  grid-template-columns: repeat(3, 375 * $px);
```
![image](https://user-images.githubusercontent.com/25207759/114409935-9f5c2700-9bb3-11eb-976f-3cf8ad6f0818.png)

sashasashasasha151 replied:
![image](https://user-images.githubusercontent.com/25207759/114410137-c9154e00-9bb3-11eb-8d50-e3a03a56d7a1.png)

sashasashasasha151 committed:
Change container and grid sizes

pva701 committed:
Preseve Plotly zoom

Problem: now when point clicked, state changed and we fully re-render chart,
as result plotly zoom is resetted.

Solution: preserve plotly layout as react state.

pva701 committed:
Fix small screen issues

Problem: when viewport is small, page is getting messy.

Solution: make it scrollable horizontally in this case.

pva701 committed:
Disable unnecessary plotly modeBar buttons

Problem: there are redundant plotly mode bar buttons.

Solution: disable them.

pva701 committed:
Rewrite new subexperiment logic. Add disabled points on the plot

Problem: it's not possible to enabled disable points, disable points
are not even shown on the chart.

Solution: change a state of new subexperiment a little bit,
add traces for disabled points, also for currently disabled and currently enabled.

pva701 committed:
Make possible to create new subexperiments

Problem: it's not possible to create new subexperiments.

Solution: show an overlay plate to control it.

pva701 committed:
Adjust new api from master

Problem: result format is changed a little bit.

Solution: adjust to new api.

pva701 committed:
Small stylistic fixes

Problem: border of last row is shown unconditionally,
also rounding for new subepxeriment name looks quirk.

Solution: fix issues.

sashasashasasha151 commented on frontend/src/App.tsx:
```diff
@@ -1,20 +1,73 @@
+    }
+  });
+
+  useEffect(() => {
+    // Cleaners of state when we leave a page
+    if (location.pathname !== "/upload") {
+      resetExcelFile();
+    }
+
+    if (location.pathname !== "/dashboard") {
+      resetColorsCounter();
+      resetSubexperimentsColors();
+      resetProjectSelectedId();
+      resetCompoundIdSelected();
+      resetTargetIdSelected();
+      resetExperimentsTableSizeAtom();
+      // For some weird reason reset doesn't update selectedExperimentsQuery selector
+      setSelectedSubExperimentsIds(new Set<number>());
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [location.pathname]);
```
`Route` has attribute `render`, so you don't need this `useEffect` with ifs. Also , why don't you load all needed information for pages here?

pva701 replied:
>Route has attribute render

and what? how can replace useEffect? it's an equivalent of children of Route. how it's helpful and replacement of useEffect? you should learn react-router better, perpahs :putya:

>why don't you load all needed information for pages here?

because there are plenty of things to finish, and I don't consider it's 100% better: you might have ton of experiments, and when you go on the website to just look at projects, you will be obliged to load all this shit. also, if you want to see information about several experiments, when you have 1000 of them, in your suggestion, I need to upload ton of irrelevant bullshit. So there are a lot of drawback in such approach and I didn't have time for pieces of information which might be reasonable o pre-fetch. Also, I would rather fix updating states: for now there are places where all the data (a list) if refetched when one item of list is updated.

sashasashasasha151 commented on frontend/src/components/Button/Button.tsx:
```diff
@@ -30,28 +32,28 @@ export function Button({
-        textButton_gray: btnStyle === "gray",
-        textButton_delete: btnStyle === "delete",
-      })}
+      className={cx(
+        [className],
+        cn(baseClassNm)({
+          active,
+          gray: btnStyle === "gray",
+          delete: btnStyle === "delete",
+          small: btnSize === "small",
+          halfRounding: type === "half-rounded",
+        })
+      )}
```
:ghc: :putya:

sashasashasasha151 commented on frontend/src/components/ContextActions/ContextActions.tsx:
```diff
@@ -1,18 +1,20 @@
 }
 
-export function ContextActions({ actions }: ContextActionsProps): React.ReactElement {
+export function ContextActions({ actions, className }: ContextActionsProps): React.ReactElement {
```
![image](https://user-images.githubusercontent.com/25207759/114728402-baf83680-9d47-11eb-85e8-fbe08dfe001b.png)

pva701 replied:
Fixed in the following PR

sashasashasasha151 commented on frontend/src/components/ContextActions/ContextItems.tsx:
```diff
@@ -0,0 +1,26 @@
+  onClick: () => void;
+}
+
+export function EditContextItem({ onClick }: ContextItemProps): React.ReactElement {
+  return (
+    <div key="edit" className="contextActions__item" onMouseDown={onClick}>
+      <EditSvg />
+      Edit
+    </div>
+  );
+}
+
+export function DeleteContextItem({ onClick }: ContextItemProps): React.ReactElement {
+  return (
+    <div key="delete" className="contextActions__item" onMouseDown={onClick}>
+      <DeleteSvg />
+      Delete
+    </div>
+  );
+}
```
Why not combine them with simple `type` in props?

pva701 replied:
indeed

pva701 replied:
But will be fixed in the next PR, because there will be more such context actions items, it will make more sense there.

sashasashasasha151 commented on frontend/src/components/CreatableSelect.tsx:
```diff
@@ -1,30 +1,50 @@
   [prop: string]: any;
 }
 
 export default function CreatableSelect<T>({
```
Am I right that it should be creatable, but now it isn't?

pva701 replied:
yes, name is outdated. I will rename

sashasashasasha151 commented on frontend/src/components/DescriptiveSelector/DescriptivePlate.scss:
```diff
@@ -0,0 +1,34 @@
+@import "src/styles/fonts";
+
+.descriptivePlate {
+  padding: (22 * $px) (22 * $px) (18 * $px) (18 * $px);
```
```suggestion
  padding: (22 * $px) (18 * $px);
```

sashasashasasha151 commented on frontend/src/components/DescriptiveSelector/DescriptivePlate.scss:
```diff
@@ -0,0 +1,34 @@
+  border-bottom: $px solid $color-gray;
+  border-left: $px solid $color-gray;
+  border-bottom-right-radius: $border-radius;
+  border-bottom-left-radius: $border-radius;
```
```suggestion
  border-radius: 0 0 0.5rem 0.5rem;
```

sashasashasasha151 commented on frontend/src/components/DescriptiveSelector/DescriptivePlate.scss:
```diff
@@ -0,0 +1,34 @@
+  padding: (22 * $px) (22 * $px) (18 * $px) (18 * $px);
+  border-right: $px solid $color-gray;
+  border-bottom: $px solid $color-gray;
+  border-left: $px solid $color-gray;
```
Combine

sashasashasasha151 commented on frontend/src/components/DescriptiveSelector/DescriptiveSelector.scss:
```diff
@@ -0,0 +1,14 @@
+.descriptiveSelector {
+  display: flex;
+  flex-direction: column;
+  width: 100%;
```
Not needed

sashasashasasha151 commented on frontend/src/components/DescriptiveSelector/DescriptiveSelector.scss:
```diff
@@ -0,0 +1,14 @@
+  //  border: $px solid $color-gray;
+  //  border-top-left-radius: $border-radius;
+  //  border-top-right-radius: $border-radius;
+  //}
```
Remove

sashasashasasha151 commented on frontend/src/components/DialogLayout/DialogLayout.scss:
```diff
@@ -6,7 +6,7 @@
   top: 0;
   left: 0;
-  z-index: 2;
+  z-index: 100;
```
Looks like 2 is enough

pva701 replied:
no, there are other elements with z-index, I don't want to think about all of them but I am sure this overlay should be over any other elements, so it's 100. it was fixed due to there is an another element with z-index greater than 2.

sashasashasasha151 commented on frontend/src/pages/dashboard/DashboardPage.tsx:
```diff
@@ -0,0 +1,79 @@
+    <PlotlyChart
+      className={className}
+      subExperiments={zip(subExperiments, colors)
+        .filter(([sub, color]) => isDefined(color) && "Right" in sub.meta.item.result)
```
In case of `Left` we should somehow show error message (But for now it is ok)

pva701 replied:
I had an idea for now: show --- for invalid IC50 (maybe in red color) and on hover show an error message. It would save us from handling of ton of errors from different subexperiments.

sashasashasasha151 commented on frontend/src/pages/dashboard/ExpandMinimizeButton/ExpandMinimizeButton.tsx:
```diff
@@ -0,0 +1,34 @@
+export function ExpandMinimizeButton({
+  targetSize,
+}: ExpandMinimizeButtonProps): React.ReactElement {
+  const [expTableSize, setExpTableSize] = useRecoilState(experimentsTableSizeAtom);
```
Okay, but why not to use simple state in parent

pva701 replied:
hm, makes sense perhaps.

sashasashasasha151 commented on frontend/src/store/types.ts:
```diff
@@ -40,4 +63,32 @@ export type ModalDialogState =
+  return "expanded";
+}
+
+export const chartColors = [
+  "#C6E294",
+  "#8E95D5",
+  "#E6D85D",
+  "#3D9C97",
+  "#C15959",
+  "#53AC62",
+  "#A26BC4",
+  "#6076E0",
+  "#374275",
+  "#BF5688",
+];
```
Don't you want to use color generator?

pva701 replied:
no, there was a relevant discussion on the slack where we agreed on predefined set of color from figma

pva701 committed:
Fix stylistic issues

Problem: there are some small issues, like:
outdated CreateSelector name repeating properties in css, etc.

Solution: fix all of them.
